server/src/main/java/org/opensearch/common/geo/PointParser.java

```diff
@@ -0,0 +1,227 @@
+ *
+ * @opensearch.internal
+ */
+public class PointParser {
+    private static final String FN_GEOJSON_TYPE = "type";
+    private static final String FV_GEOJSON_TYPE_POINT = "Point";
+    private static final String FN_GEOJSON_COORDS = "coordinates";
+
+    private static final String FN_GEOHASH = "geohash";
+    private static final String ERR_MSG_INVALID_TOKEN = "token [{}] not allowed";
+
+    private String fieldNameX;
+    private String fieldNameY;
+    private boolean supportGeoHash;
+
+    private String invalidFieldErrMsg;
+
+    public PointParser(final String fieldNameX, final String fieldNameY, final boolean supportGeoHash) {
+        this.fieldNameX = fieldNameX;
+        this.fieldNameY = fieldNameY;
+        this.supportGeoHash = supportGeoHash;
+        this.invalidFieldErrMsg = supportGeoHash
+            ? String.format(
+                Locale.ROOT,
+                "field must be either %s/%s, %s/%s, or %s",
+                fieldNameX,
+                fieldNameY,
+                FN_GEOJSON_TYPE,
+                FN_GEOJSON_COORDS,
+                FN_GEOHASH
+            )
+            : String.format(
+                Locale.ROOT,
+                "field must be either %s/%s, or %s/%s",
+                fieldNameX,
+                fieldNameY,
+                FN_GEOJSON_TYPE,
+                FN_GEOJSON_COORDS
+            );
+    }
+
+    public Point parseObject(final XContentParser parser, final boolean ignoreZValue, final GeoUtils.EffectivePoint effectivePoint)
+        throws IOException {
+        if (parser.currentToken() != XContentParser.Token.START_OBJECT) {
+            throw new OpenSearchParseException("object is expected");
+        }
+
+        parser.nextToken();
+
+        if (parser.currentToken() == XContentParser.Token.END_OBJECT) {
+            throw new OpenSearchParseException(invalidFieldErrMsg);
+        }
+
+        if (parser.currentToken() != XContentParser.Token.FIELD_NAME) {
+            throw new OpenSearchParseException(ERR_MSG_INVALID_TOKEN, parser.currentToken());
+        }
+
+        Point point = null;
+        String field = parser.currentName();
+        if (fieldNameX.equals(field) || fieldNameY.equals(field)) {
+            point = parseBasicFields(parser);
+        } else if (supportGeoHash && FN_GEOHASH.equals(field)) {
+            point = parseGeoHashFields(parser, effectivePoint);
+        } else if (FN_GEOJSON_TYPE.equals(field) || FN_GEOJSON_COORDS.equals(field)) {
+            point = parseGeoJsonFields(parser, ignoreZValue);
+        } else {
+            throw new OpenSearchParseException(invalidFieldErrMsg);
+        }
+
+        if (parser.currentToken() != XContentParser.Token.END_OBJECT) {
+            throw new OpenSearchParseException(invalidFieldErrMsg);
+        }
+
+        return point;
+    }
+
+    private Point parseGeoHashFields(final XContentParser parser, final GeoUtils.EffectivePoint effectivePoint) throws IOException {
```

Comment:
Given that XYPoint will also use this class to parse the point, we should keep all the functions which are related to GeoPoint only back to GeoUtils to preserve the cohesion of the class.

Reply:
Same as above. If we want to move this away from PointParser, the `parseObject` logic should be implemented in both GeoUtils and XYPointParser.

Reply:
make the changes as suggested above.